lib/puppetserver/ca/revoke_action.rb

```diff
@@ -0,0 +1,89 @@
+require 'optparse'
+
+module Puppetserver
+  module Ca
+    class RevokeAction
+      # Actions require a short summary constant
+      SUMMARY = 'Revoke a given certificate'
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca revoke [--help|--version]
+  puppetserver ca revoke CERTNAME
+
+Description:
+Given a valid certname, instructs the CA to revoke it.
+
+Options:
+BANNER
+
+      # class method parser returns an instance of Ruby's OptionParser
+      # (or a custom object that implements summarize() and help().
+      def self.parser(parsed = {})
+        OptionParser.new do |o|
+          o.banner = BANNER
+          o.on('--help', 'Display this revoke specific help output') do |help|
+            parsed['help'] = true
+          end
+          o.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      # take unparsed array of arguments, return parsed hash and optionally an error code
+      def parse(args)
+        results = { certname: [] }
+        unparsed = []
+        parser = self.class.parser(results)
+
+        begin
+          parser.order!(args) do |unknown|
+            results[:certname] << unknown
+          end
+        rescue OptionParser::ParseError => e
+          unparsed += e.args
+          unparsed << args.shift unless args.first =~ /^-{1,2}/
+          retry
+        end
+
+        exit_code = nil
+        unless unparsed.empty?
+          @logger.err 'Error:'
+          @logger.err 'Unknown arguments or flags:'
+          unparsed.each do |arg|
+            @logger.err "    #{arg}"
+          end
+
+          @logger.err ''
+          @logger.err parser.help
+          exit_code = 1
```

Comment:
This also might be good in a util method along with the unknown option parsing above? Or we could try to refactor it into the top level Cli class so that actions don't need to know about it??

Reply:
I think it makes sense to keep it in the actions, but a utility method definitely seems good.